src/test/peerman_tests.cpp

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <chainparams.h>
+#include <node/miner.h>
+#include <net_processing.h>
+#include <pow.h>
+#include <test/util/setup_common.h>
+#include <validation.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(peerman_tests, RegTestingSetup)
+
+// Verifying when network-limited peer connections are desirable based on the node's IBD status
+BOOST_AUTO_TEST_CASE(connections_desirable_service_flags)
+{
+    std::unique_ptr<PeerManager> peerman = PeerManager::make(*m_node.connman, *m_node.addrman, nullptr, *m_node.chainman, *m_node.mempool, {});
+
+    // Check we start connecting to full nodes
+    ServiceFlags peer_flags{NODE_WITNESS | NODE_NETWORK_LIMITED};
+    BOOST_CHECK(peerman->GetDesirableServiceFlags(peer_flags) == ServiceFlags(NODE_NETWORK | NODE_WITNESS));
+
+    // Check we accept network limited peers when we are out of ibd
+    SetMockTime(WITH_LOCK(::cs_main, return m_node.chainman->ActiveChain().Tip()->GetBlockTime())); // time to tip time so we are out of ibd
+    peerman->CheckForStaleTipAndEvictPeers();  // Update ibd state
+    BOOST_CHECK(peerman->GetDesirableServiceFlags(peer_flags) == ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS));
+
+    // Now check we can go back to ibd if we are further than the limited peers threshold (NODE_NETWORK_LIMITED_MIN_BLOCKS)
+    auto consensus = m_node.chainman->GetParams().GetConsensus();
+    auto time_last_stale = GetTime<std::chrono::seconds>() + std::chrono::seconds{consensus.nPowTargetSpacing * 300};
+    SetMockTime(time_last_stale);
+    peerman->CheckForStaleTipAndEvictPeers();
+    BOOST_CHECK(peerman->GetDesirableServiceFlags(peer_flags) == ServiceFlags(NODE_NETWORK | NODE_WITNESS));
+
+    // Check we can get back out of ibd again
+    SetMockTime(WITH_LOCK(::cs_main, return m_node.chainman->ActiveChain().Tip()->GetBlockTime())); // time to tip time so we are out of ibd
+    peerman->CheckForStaleTipAndEvictPeers();  // Update ibd state
+    BOOST_CHECK(peerman->GetDesirableServiceFlags(peer_flags) == ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS));
+
+    // Push the chain forward to surpass the 'STALE_CHECK_INTERVAL' (time between 'CheckForStaleTipAndEvictPeers' calls)
+    // to move the "initial sync finished" state flag back to false (back to IBD state once more).
```

Comment:
The 3 lines below are the same as lines 38-40, except for the +11min. It's unclear to me from this comment what the 11min is doing and why this is different from what we are doing before. Can this maybe be explained better?

Reply:
Sure. This is because we execute the stale check every 10 min, so needed to push the last stale check time forward.
Will re-order the test checks so this becomes clearer.